### PR TITLE
Guest Post: Monitoring Dreamhack - the World's Largest Digital Festival

### PR DESCRIPTION
**Pull request by bluecmd**
See commits for the guest story of how we used Prometheus to monitor Dreamhack.

When I send this PR I still have one image to replace, but other than that: RFC.

Ping @juliusv.

### CODE REVIEW

File: content/blog/2015-06-20-monitoring-dreamhack.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+[![dhmon Architecture](http://i.imgur.com/6gN3MRp.png)](http://i.imgur.com/6gN3MRp.png)
+<center>*The current architecture plan of dhmon as of Summer 2015*</center>
+
+As for storage we use Prometheus as a central timeseries storage and querying engine but we also use Redis and memcache to export snapshot views of binary information that we collect but cannot store in Prometheus in any sensible way or when we need to access very fresh data.
+
+One such case is in our presentation layer. We use our dhmap web application to get an overview of the overall health of the access switches. In order to be effective at resolving errors we need a latency of ~10 seconds from data collection to presentation. Our goal is to have fixed the problem before the customer notices, or at least before he/she has walked over to the support people to report an issue. For this reason, we have been using memcache since the beginning to access the latest snapshot of the network.
```
Comment (fabxc) [inline content/blog/2015-06-20-monitoring-dreamhack.md]: If we are referring to the caching server, it's correct name is memchache*d*.

File: content/blog/2015-06-20-monitoring-dreamhack.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+## Prometheus Setup
+The block that so far has been referred to as Prometheus really consists of three products: Prometheus, PromDash and Alertmanager. The setup is fairly basic and all three softwares are running on the same host. Everything is served by an Apache web server that just acts as a reverse proxy.
+
+    ProxyPass /prometheus http://localhost:9090/prometheus
+    ProxyPass /alertmanager http://localhost:9093/alertmanager
+    ProxyPass /dash http://localhost:3000/dash
+
+## Exploring the Network
+Prometheus is a competent querying engine, and powered by streaming information from all over your network allows you to do pretty cool things. However sometimes the queries need to process too much data to finish within a reasonable amount of time. This happened to us when we wanted to graph the top 5 utilized links. While the query worked it would take about the time we set our timeout limit to, meaning it was both slow and flaky. We decided to try to use Prometheus' precomputing facility.
```
Comment (fabxc) [inline content/blog/2015-06-20-monitoring-dreamhack.md]: I had a hard time reading the first sentence. Maybe:

Prometheus *has* a competent querying engine *that*, powered by streaming information from all over your network, allows you to do pretty cool things.

**Comment by fabxc** [comment]
In other Markdown files we are breaking lines manually. In general I find wrapping by the editor way better as otherwise average line length varies across files and changes require tedious reformatting. But we should probably be consistent with what we have so far – or do one big search/replace.

If it is possible license-wise, I'd vote for self-hosting the images. Right now it seems like flickr and imgur will stick around forever but this blog might survive them ;)

Great read! Thanks for the effort!

File: content/blog/2015-06-20-monitoring-dreamhack.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+For the more major features we're planning on adding: syslog analysis (we have a lot of syslog!), alerts from our intrusion detection systems, integration with our Puppet setup, and also integrate more across the different teams at Dreamhack. We managed to create a proof-of-concept where we got data from one of the electrical current sensors into our monitoring, making it easy to see if a device is faulty or if it simply don't have any electricity anymore. We're also working on integrating to the Point-of-Sale systems that are used in the stores at the event. Who doesn't want to graph the sales of ice cream?
+
+Finally not all services that the team operate are on-site and some even run 24/7, even after the event. We want to monitor these services with Prometheus as well, and in the long run when Prometheus gets support for federation utilize the off-site Prometheus to replicate the metrics from the event Prometheus.
```
Comment (fabxc) [inline content/blog/2015-06-20-monitoring-dreamhack.md]: Finally,

**Comment by juliusv** [comment]
FYI, chatted with @bluecmd - I'm doing an editorial wording pass over this. Feel free to comment content-wise, but ignore wording / spelling / grammar until after that first pass please.

**Comment by juliusv** [comment]
My rewording PR, which funnels into this one, is here: https://github.com/bluecmd/prometheus-docs/pull/1

I included @fabxc's comments, except for the images (don't feel strongly about that) and the word-wrapping. I would leave the word wrapping as the last step so we don't drown in diff noise in the meantime.

**Comment by bluecmd** [comment]
Thanks for the help! Looks awesome. I added a commit with the final Dashboard screenshot as well now.
As for mirroring the images, I'd rather not due to copyrighting by Dreamhack. Linking to the Flickr stream seems to be the most straight forward way.

Unless I'm missing something, that should leave only the word-wrapping. Let me know if you have any other concerns or I'll wrap it to 80 char lines.

**Comment by fabxc** [comment]
Great!

As the images are no info graphics or anything that are required to understand the content I'm fine with leaving it as is.

**Comment by grobie** [comment]
Great article!

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+created_at: 2015-06-24
+kind: article
+author_name: Christian Svensson (DreamHack Network Team)
+---
+
+*Editor's note: This article is a guest post written by a Prometheus user.*
+
+**If you are operating the network for 10,000's of demanding gamers, you need to really know what is going on inside your network. Oh, and everything needs to be built from scratch in just five days.**
+
+If you have never heard about DreamHack before, here is the pitch:
```
Comment (grobie) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: Should "DreamHack" be a link to the dreamhack website?

Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: Oh yes please.

Comment (bluecmd) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: Good idea, done!

**Comment by bluecmd** [comment]
There we go, all wrapped where it makes sense.

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+computers.
+
+[![Access switches](https://c1.staticflickr.com/9/8487/8206439882_4739d39a9c_c.jpg)](https://www.flickr.com/photos/dreamhack/8206439882)
+<center>*Dutifully standing in line, the access switches are ready to greet the
+DreamHackers with high-speed connectivity.*</center>
+
+Obviously just connecting all these computers to a switch is not enough. That
+switch needs to be connected to the other switches as well. This is where the
+distribution switches (or dist switches) come into play. These are switches that
+take the hundreds of links from all access switches and aggregates them into
```
Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: s/aggregates/aggregate/

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+[![Network planning map](http://i.imgur.com/ZCQa2Abl.png)](http://i.imgur.com/ZCQa2Ab.png)
+<center>*The planning map for the distribution and core layers. The core is
+clearly visible in "Hall D"*</center>
+
+All in all this is becoming a lengthy list of stuff to monitor, so let's get to
+the reason you're here: How do we make sure we know what's going on?
+
+## Introducing: dhmon
+dhmon is the collective name of the systems that not only
+monitor the network, but also allows other teams to collect metrics on whatever
```
Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: s/allows/allow/

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+they want.
+
+Since the network needs to be built in five days, it's essential that the
+monitoring systems are easy to set up and keep in sync if we need to do last
+minute infrastructural changes (like adding or removing devices). When we start
+to build the network, we need monitoring as soon as possible to be able to
+discover any problems with the equipment or other issues we hadn't foreseen.
+
+In the past we have tried to use a mix of commonly available software such as
+Cacti, SNMPc, Opsview among others. While these have worked they have focused on
```
Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: ", Opsview" -> ", and Opsview"

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+After this, running `topk(5, precomputed_link_utilization_percent)` was
+blazingly fast.
+
+## Being reactive: alerting
+So at this stage we had something we could query for
+the state of the network. Since we are humans, we don't want to spend our time
+running queries all the time to see if things are still running as they should,
+so obviously we need alerting.
+
+For example: we know that all our access switches uses GigabitEthernet0/2 as an
```
Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: "uses" -> "use"

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+this OID is indexed by an arbitrary interface index. To make any sense of this
+index, we need to cross-reference it with data from SNMP OID `IF-MIB::ifDescr`
+to retrieve the actual interface name.
+
+Fortunately, our snmpcollector supports this kind of cross-referencing while
+generating Prometheus metrics. This allows us in a simple way to not only query
+data, but also define useful alerts. In our setup we configured the SNMP
+collection to annotate any metric under the `IF-MIB::ifTable` and
+`IF-MIB::ifXTable` OIDs with `ifDescr`. This will come in handy now when we need
+to specify that we are only interested in `GigabitEthernet0/2` port and no other
```
Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: "in" -> "in the" (my bad)

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+<center>*The DreamHack Overview dashboard powered by PromDash*</center>
+
+## The future
+While changing an integral part of any system is a complex job and
+we're happy that we managed to integrate Prometheus in just one event, there are
+without a doubt a lot of areas to improve. Some areas are pretty basic: using
+more precomputed metrics to improve performance, adding more alerts, and tuning
+the ones we have. Another area is to make it easier for operators: creating an
+alert dashboard suitable for our network operations center (NOC), figuring out
+if we want to page the people on-call, or just letting the NOC escalate alerts.
```
Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: "letting" -> "let"

File: content/blog/2015-06-24-monitoring-dreamhack.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+Some bigger features we're planning on adding: syslog analysis (we have a lot of
+syslog!), alerts from our intrusion detection systems, integrating with our
+Puppet setup, and also integrating more across the different teams at DreamHack.
+We managed to create a proof-of-concept where we got data from one of the
+electrical current sensors into our monitoring, making it easy to see if a
+device is faulty or if it simply doesn't have any electricity anymore. We're
+also working on integrating with the point-of-sale systems that are used in the
+stores at the event. Who doesn't want to graph the sales of ice cream?
+
+Finally, not all services that the team operate are on-site, and some even run
```
Comment (juliusv) [inline content/blog/2015-06-24-monitoring-dreamhack.md]: "operate" -> "operates"

**Comment by juliusv** [comment]
I made a last round of tiny comments. After that, :+1:. Maybe even squash the whole thing (don't care if my commit is squashed away).

**Comment by bluecmd** [comment]
Fixed and squished!

**Comment by juliusv** [comment]
:+1: Woot!

**Pull request merged**